---
title: "Class 9: Structural Bioinformatics pt 1"
author: "Seona Patel (PID: A69035519)"
format:
  pdf:
    fig-width: 8
    fig-height: 6
---

## The PDB Database

The main repository for biomolecular structure data is the Protein Data Bank (PDB) https://www.rcsb.org

Let's have a quick look at the composition of this database: 

```{r}
stats <- read.csv("Data Export Summary.csv")
stats
```

> Q1: What percentage of structures in the PDB are solved by X-Ray and Electron Microscopy.

```{r}
stats$X.ray <- gsub(",", "", stats$X.ray)
sum(as.numeric(stats$X.ray))

```

This is annoying, lets try a different import function from the **readr** package. 

```{r}
library(readr)
```
```{r}
stats <- read_csv("Data Export Summary.csv")
stats
```
```{r}
round(sum(stats$EM)/sum(stats$Total)*100,2)
```

```{r}
round(sum(stats$`X-ray`)/sum(stats$Total)*100,2)
```

**So 81.43% of the structures in the PDB are solved by X-ray and 12.27% of the structures are solved by EM.** 


> Q2: What proportion of structures in the PDB are protein?

```{r}
round(stats[1,9]/sum(stats$Total)*100,2)

```
So 86.05% are protein.

> Q3: Make a bar plot overview of Molecular type composition using ggplot

```{r}
library(ggplot2)
library(tidyr)
```


```{r}
stats$`Molecular Type` <- factor(stats$`Molecular Type`, levels = c("Protein (only)", "Protein/Oligosaccharide", "Protein/NA", "Nucleic Acid (only)", "Oligosaccharide (only)","Other"))

```

```{r}
stats_long <- pivot_longer(
  stats,
  cols = -c(Total, `Molecular Type`),
  names_to = "ExperimentalType",
  values_to = "Count"
)
head(stats_long)
```
```{r}
ggplot(stats_long) +
  aes(x=`Molecular Type`, y = Count, fill = ExperimentalType) +
  geom_col() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) 
  
```
```{r}
library(dplyr)
totals_df <- stats_long |>
  group_by(`Molecular Type`) |>
  summarise(Total = sum(Count))
```


```{r}
ggplot(stats_long) +
  aes(x = `Molecular Type`, y = Count, fill = ExperimentalType) +
  geom_col() +
  geom_text(
    data = totals_df,
    aes(x = `Molecular Type`, y = Total, label = Total),
    vjust = -0.5,
    inherit.aes = FALSE
  )+
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
```

>Q4: Water molecules normally have 3 atoms. Why do we see just one atom per water molecule in this structure?

We only see one oxygen per water molecule because hydrogen atoms are hard to detect due to their low electron density in x-ray crystallography. 

>Q5: There is a critical “conserved” water molecule in the binding site. Can you identify this water molecule? What residue number does this water molecule have

This water molecule is HOH 308.

## Visualizing structure data

The Mol* viewer is embedded in many bioinformatics websites.The homepage is https://molstar.org/


I can insert any figure or image file using markdown format. 

> Q6: Generate and save a figure clearly showing the two distinct chains of HIV-protease along with the ligand. You might also consider showing the catalytic residues ASP 25 in each chain and the critical water (we recommend “Ball & Stick” for these side-chains). Add this figure to your Quarto document.


![The HIV-Pr dimer with bound inhibitor](1HSG.png)


![The catalytic ASP25 and active-site water](1HSG_water.png)


##Bio3D package for structural bioinformatics

We can use the bio3d package to read and analyze biomolecular data in R: 

```{r}
library(bio3d)

hiv <- read.pdb(file = "1hsg")
hiv
```
>Q7: How many amino acid residues are there in this pdb object? 

198 amino acids

>Q8: Name one of the two non-protein residues? 

MK1 

>Q9: How many protein chains are in this structure? 

There are 2 protein chains in this structure

```{r}
head(hiv$atom)
```
```{r}
pdbseq(hiv)
```

Let's trim to chain A and get just it's sequence

```{r}
chainA <- trim.pdb(hiv, chain="A")
chainA.seq <- pdbseq(chainA)
chainA.seq
```

Let's blast 

**Cache the blast so rendering doesn't take forever**

```{r blast_hiv, cache=TRUE}
blast <- blast.pdb(chainA.seq)

```

```{r}
head(blast$hit.tbl)
```

Plot a quick overview of blast results 

```{r}
hits <- plot(blast)
```


Every dot is a hit on this. 

Get the accession number of all of the hits:

```{r}
hits$pdb.id
```


## Prediction of Functional motions

We can run an Normal Mode Analysis (NMA) to predict large scale motions/flexibility/dynamics of any biomolecule that we can read into R. 


Let's look at ADK and chain A only!


```{r}
adk <- read.pdb("1ake")
adk_A <- trim.pdb(adk, chain="A")
adk_A
```

```{r}
m <- nma(adk_A)
plot(m)
```
This tells me about the flexibility of the protein.

Let's write out a "trajectory" of predicted motion

```{r}
mktrj(m, file="adk_nma.pdb")
```

## Play with 3D viewing in R

We can use the new **bio3dview** package, which is not yet on CRAN, to render interactvie 3D views in R and HTML quarto output reports. 


To install from GitHub we can use the **pak** package.

```{r}
pak::pak("bioboot/bio3dview")
```

```{r}
library(bio3dview)
#view.pdb(adk)
```

## Comparative structure analysis of Adenylate Kinase

The goal of this section is to perform principal component analysis (PCA) on the complete collection of Adenylate kinase structures in the protein data-bank (PDB).

**Adenylate kinase** (often called simply Adk) is a ubiquitous enzyme that functions to maintain the equilibrium between cytoplasmic nucleotides essential for many cellular processes. Adk operates by catalyzing the reversible transfer of a phosphoryl group from ATP to AMP. This reaction requires a rate limiting conformational transition (i.e. change in shape). Here we analyze all currently available Adk structures in the PDB to reveal detailed features and mechanistic principles of these essential shape changing transitions.


The bio3d package pca() function provides a convenient interface for performing PCA of biomolecular structure data. As we have discussed in previous classes, PCA is a statistical approach used to transform large data-sets down to a few important components that usefully describe the directions where there is most variance. In terms of protein structures PCA can be used to capture major structural variations within a set of structures (a.k.a. structure ensemble). This can make interpreting major conformational states (such as ‘active’ and ‘inactive’ or ‘ligand bound’ and ‘un-bound’ states) and structural mechanisms for activation or regulation more clear.

## Overview

Starting from only one Adk PDB identifier (PDB ID: 1AKE) we will search the entire PDB for related structures using BLAST, fetch, align and superpose the identified structures, perform PCA and finally calculate the normal modes of each individual structure in order to probe for potential differences in structural flexibility.

> Q10. Which of the packages above is found only on BioConductor and not CRAN? 

msa

> Q11. Which of the above packages is not found on BioConductor or CRAN?: 

bio3d-view

> Q12. True or False? Functions from the devtools package can be used to install packages from GitHub and BitBucket? 

TRUE

## Search and retrieve ADK structures 

Below we perform a blast search of the PDB database to identify related structures to our query Adenylate kinase(ADK) sequence. In this particular example we use function get.seq() to fetch the query sequence for chain A of the PDB ID 1AKE and use this as input to blast.pdb(). Note that get.seq() would also allow the corresponding UniProt identifier.

```{r}
library(bio3d)
aa <- get.seq("1ake_A")
```

```{r}
aa
```

>Q13. How many amino acids are in this sequence, i.e. how long is this sequence? 

214

Now we can use this sequence as a query to BLAST search the PDB to find similar sequences and structures.

```{r blast_adk, cache=TRUE}
b <- blast.pdb(aa)
```


The function plot.blast() facilitates the visualization and filtering of the Blast results. It will attempt to set a seed position to the point of largest drop-off in normalized scores (i.e. the biggest jump in E-values). In this particular case we specify a cutoff (after initial plotting) of to include only the relevant E.coli structures:

```{r}
# Plot a summary of search results 
hits <- plot(b)
```

```{r}
# List out some 'top hits'
head(hits$pdb.id)
```

The Blast search and subsequent filtering identified a total of 13 related PDB structures to our query sequence. The PDB identifiers of this collection are accessible through the `$pdb.id` attribute to the hits object (i.e. `hits$pdb.id`). Note that adjusting the cutoff argument (to `plot.blast()`) will result in a decrease or increase of hits.

We can now use function `get.pdb()` and `pdbslit()` to fetch and parse the identified structures.

```{r}
# Download releated PDB files
files <- get.pdb(hits$pdb.id, path="pdbs", split=TRUE, gzip=TRUE)
```
## Align and superpose structures

Next we will use the `pdbaln()` function to align and also optionally fit (i.e. superpose) the identified PDB structures.

```{r}
# Align related PDBs
pdbs <- pdbaln(files, fit = TRUE, exefile="msa")
```


```{r}
# Vector containing PDB codes for figure axis
ids <- basename.pdb(pdbs$id)

# Draw schematic alignment

plot(pdbs, labels=ids)
```
Figure 7: Schematic representation of alignment. Grey regions depict aligned residues, while white depict gap regions. The red bar at the top depict sequence conservation

## Annotate collected PDB structures

The function pdb.annotate() provides a convenient way of annotating the PDB files we have collected. Below we use the function to annotate each structure to its source species. This will come in handy when annotating plots later on:

```{r}
anno <- pdb.annotate(ids)
unique(anno$source)
```
We can view all available annotation data:

```{r}
anno
```


## Principle Component Analysis

Function pca() provides principal component analysis (PCA) of the structure data. PCA is a statistical approach used to transform a data set down to a few important components that describe the directions where there is most variance. In terms of protein structures PCA is used to capture major structural variations within an ensemble of structures.

PCA can be performed on the structural ensemble (stored in the pdbs object) with the function pca.xyz(), or more simply pca().

```{r}
# Perform PCA
pc.xray <- pca(pdbs)
plot(pc.xray)
```
Figure 9: Results of PCA on Adenylate kinase X-ray structures. Each dot represents one PDB structure.

Function rmsd() will calculate all pairwise RMSD values of the structural ensemble. This facilitates clustering analysis based on the pairwise structural deviation:

```{r}
# Calculate RMSD
rd <- rmsd(pdbs)

# Structure-based clustering
hc.rd <- hclust(dist(rd))
grps.rd <- cutree(hc.rd, k=3)

plot(pc.xray, 1:2, col="grey50", bg=grps.rd, pch=21, cex=1)
```
Figure 10: Projection of Adenylate kinase X-ray structures. Each dot represents one PDB structure.

The plot shows a conformer plot – a low-dimensional representation of the conformational variability within the ensemble of PDB structures. The plot is obtained by projecting the individual structures onto two selected PCs (e.g. PC-1 and PC-2). These projections display the inter-conformer relationship in terms of the conformational differences described by the selected PCs.


## 5. Optional further visualization

To visualize the major structural variations in the ensemble the function mktrj() can be used to generate a trajectory PDB file by interpolating along a give PC (eigenvector):

```{r}
# Visualize first principal component
pc1 <- mktrj(pc.xray, pc=1, file="pc_1.pdb")
```


```{r}
#Plotting results with ggplot2
library(ggplot2)
library(ggrepel)

df <- data.frame(PC1=pc.xray$z[,1], 
                 PC2=pc.xray$z[,2], 
                 col=as.factor(grps.rd),
                 ids=ids)

p <- ggplot(df) + 
  aes(PC1, PC2, col=col, label=ids) +
  geom_point(size=2) +
  geom_text_repel(max.overlaps = 20) +
  theme(legend.position = "none")
p
```


# Comparative analysis of protein structures

Starting wth a sequence or structure ID (accession number) let's run a complete analysis pipeline.

```{r}
library(bio3d)
id <- "1ake_A"
aa <- get.seq(id)
aa
```
```{r, cache=TRUE}
blast <- blast.pdb(aa)
hits <- plot(blast)
```
```{r}
hits$pdb.id
```

Download all of these "hits" that are similar to our starting id sequence
```{r}
# Download releated PDB files
files <- get.pdb(hits$pdb.id, path="pdbs", split=TRUE, gzip=TRUE)
```
Now we will align and superimpose these structures

```{r}
# Align releated PDBs
pdbs <- pdbaln(files, fit = TRUE, exefile="msa")

```


```{r}
pc.xray <- pca(pdbs)
plot(pc.xray) 

```

```{r}
plot(pc.xray, 1:2)
```
```{r}
mktrj(pc.xray, file="pca_results.pdb")
```

```{r}

library(bio3dview)
view.pca(pc.xray)
```

